app/routes.py

```diff
@@ -1,2 +1,219 @@
+        all_tasks.append(Task.query.get(task_id))
+
+    goal.tasks = all_tasks
+
+    db.session.commit()
+
+    new_response = {
+    "id": goal.id,
+    "task_ids": request_body["task_ids"]
+    }
+
+    return jsonify(new_response), 200
+
+@goals_bp.route("", methods=["POST"])
+def post_goals():
+    request_body = request.get_json()
+
+    new_goal = Goal.from_dict(request_body)
+
+    if "title" not in request_body:
+        return ({
+        "details": "Invalid data"
+    }), 400
+
+    db.session.add(new_goal)
+    db.session.commit()
+
+    return jsonify({"goal" : new_goal.to_dict()}), 201
+
+@tasks_bp.route("", methods=["POST"])
+def post():
+    request_body = request.get_json()
+
+    new_task = Task.from_dict(request_body)
+
+    if "title" not in request_body or "description" not in request_body or "completed_at" not in request_body:
+        return ({
+        "details": "Invalid data"
+    }), 400
+
+    db.session.add(new_task)
+    db.session.commit()
+
+    return jsonify({"task" : new_task.to_dict()}), 201
+
+def sort_filter_tasks():
+    sort_query = request.args.get("sort")
+    if sort_query == "desc":
+        tasks = Task.query.order_by(Task.title.desc())
+    elif sort_query == "asc":
+        tasks = Task.query.order_by(Task.title.asc())
+    else:
+        tasks = Task.query.all()
+    
+    filter_query = request.args.get("title")
+    if filter_query:
+        tasks = Task.query.filter(Task.title.contains(filter_query))
+    
+    return tasks
+
+@tasks_bp.route("", methods=["GET"])
+def get_tasks():
+    tasks = sort_filter_tasks()
+
+    tasks_response = [task.to_dict() for task in tasks]
+
+    return jsonify(tasks_response), 200
+
+def sort_filter_goals():
+    sort_query = request.args.get("sort")
+    if sort_query == "desc":
+        goals = Goal.query.order_by(Goal.title.desc())
+    elif sort_query == "asc":
+        goals = Goal.query.order_by(Goal.title.asc())
+    else:
+        goals = Goal.query.all()
+
+    filter_query = request.args.get("title")
+    if filter_query:
+        goals = Goal.query.filter(Goal.title.contains(filter_query))
+    
+    return goals
+
+@goals_bp.route("", methods=["GET"])
+def get_goals():
+    goals = sort_filter_goals()
+
+    goal_response = [goal.to_dict() for goal in goals]
+
+    return jsonify(goal_response), 200
+
+@goals_bp.route("/<goal_id>", methods=["GET"])
+@require_goal
+def get(goal):  
+    return jsonify({"goal": goal.to_dict()}), 200
+
+@tasks_bp.route("/<task_id>", methods=["GET"])
+@require_task
+def get(task):  
+    if task.goal_id:
+        return jsonify({"task": task.task_to_dict_w_goal()}), 200
+    return jsonify({"task": task.to_dict()}), 200
+
+@goals_bp.route("/<goal_id>/tasks", methods=["GET"])
+@require_goal
+def get_tasked_goal(goal):
+    return jsonify(goal.goal_task_dict()), 200
+
+@goals_bp.route("/<goal_id>", methods=["PUT"])
+@require_goal
+def put(goal):
+    form_data = request.get_json()
+
+    goal.replace_with_dict(form_data)
```

Comment:
The logic in this helper at  least protects the application from crashing if there happen to be missing fields in the input, but recall that PUT implies replacing an entire record, while PATCH would be more appropriate for "if the value is present, update it". So we should have similar strong checks for basically the same fields that we use for creating a new record applied to the logic for performing a PUT.